File: frontend/types/analysis.ts
/** The six authorship profiles the engine measures. */
export type ProfileKey =
  | "lexical"
  | "syntactic"
  | "grammatical"
  | "mechanical"
  | "stylistic"
  | "discourse";

/**
 * How trustworthy a single measurement is.
 *
 * "approximation" features are proxies with documented limits - the UI
 * badges them so a teacher doesn't read them as hard evidence.
 */
export type Measurement = "direct" | "approximation";

export type FeatureKind = "scalar" | "count" | "distribution";

export interface FeatureBreakdown {
  key: string;
  label: string;
  kind: FeatureKind;
  measurement: Measurement;
  weight: number;
  note: string;
  /** False when the submission was too short (or the wrong format) to measure this. */
  available: boolean;
  suppressed_reason: string | null;
  z: number;
  score: number | null;
  submission: number | null;
  baseline_mean: number | null;
  baseline_stdev: number | null;
  baseline_n: number | null;
}

export interface ProfileBreakdown {
  label: string;
  /** Null when every feature in the profile was suppressed. */
  score: number | null;
  z: number | null;
  weight: number;
  available: boolean;
  suppressed_reason: string | null;
  features: FeatureBreakdown[];
}

export interface OverallScore {
  z: number;
  score: number;
}

/**
 * How much weight this particular comparison can bear.
 *
 * Distinct from `confidence_level`, which describes the baseline. Keeping
 * them apart is what stops a 92% on a 120-word paragraph from looking as
 * solid as a 92% on a 900-word essay.
 */
export interface Reliability {
  confidence_level: number;
  n_baseline_papers: number;
  total_baseline_words: number;
  submission_word_count: number;
  paragraphs_reliable: boolean;
}

export interface AnalysisBreakdown {
  schema_version: number;
  extractor_version: string;
  overall: OverallScore;
  profiles: Record<ProfileKey, ProfileBreakdown>;
  reliability: Reliability;
  threshold: number;
  /** Computed server-side - the client never infers a verdict itself. */
  flagged: boolean;
  explanation: string;
}

export interface AnalysisResult {
  id: number;
  paper_id: number;
  consistency_score: number;
  confidence_level: number;
  flagged: boolean;
  breakdown: AnalysisBreakdown;
  explanation: string;
}

export type FeedbackDecision = "genuine" | "flagged";

export interface FeedbackRequest {
  decision: FeedbackDecision;
  /** The API accepts `str | null`; null means "no remark", not an empty one. */
  remarks?: string | null;
}

export interface Feedback {
  id: number;
  paper_id: number;
  decision: FeedbackDecision;
  remarks: string | null;
  created_at: string;
}

/** Display order, strongest authorship signal first. */
export const PROFILE_ORDER: ProfileKey[] = [
  "stylistic",
  "syntactic",
  "lexical",
  "mechanical",
  "discourse",
  "grammatical",
];

File: backend/schemas/analysis.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

# Every model here sets extra="forbid". The previous version relied on
# Pydantic's default extra="ignore" and unpacked the stored JSON with
# `BreakdownScore(**data)`, which silently swallowed three keys. That worked
# by accident: any future key rename would have failed silently rather than
# loudly. Shape drift should break the boundary, not pass through it.
_STRICT = ConfigDict(extra="forbid")


class FeatureBreakdown(BaseModel):
    """One measured feature, with the evidence behind its score."""

    key: str
    label: str
    kind: str
    measurement: str
    weight: float
    note: str = ""
    available: bool
    suppressed_reason: str | None = None
    z: float
    score: float | None = None
    submission: float | None = None
    baseline_mean: float | None = None
    baseline_stdev: float | None = None
    baseline_n: int | None = None

    model_config = _STRICT


class ProfileBreakdown(BaseModel):
    """One of the six authorship profiles."""

    label: str
    score: float | None
    z: float | None
    weight: float
    available: bool
    suppressed_reason: str | None = None
    features: list[FeatureBreakdown] = Field(default_factory=list)

    model_config = _STRICT


class OverallScore(BaseModel):
    z: float
    score: float

    model_config = _STRICT


class Reliability(BaseModel):
    """How much weight this particular comparison can bear.

    Distinct from `confidence_level`, which describes the baseline. Keeping
    them apart is what stops a 92% on a 120-word paragraph from looking as
    solid as a 92% on a 900-word essay.
    """

    confidence_level: float
    n_baseline_papers: int
    total_baseline_words: int
    submission_word_count: int
    paragraphs_reliable: bool

    model_config = _STRICT


class AnalysisBreakdown(BaseModel):
    schema_version: int
    extractor_version: str
    overall: OverallScore
    profiles: dict[str, ProfileBreakdown]
    reliability: Reliability
    threshold: float
    # Computed server-side so the client never has to infer a verdict by
    # comparing a score against a threshold itself.
    flagged: bool
    explanation: str

    model_config = _STRICT


class AnalysisResultResponse(BaseModel):
    id: int
    paper_id: int
    consistency_score: float
    confidence_level: float
    flagged: bool
    breakdown: AnalysisBreakdown
    explanation: str

    model_config = ConfigDict(from_attributes=True)


class FeedbackCreate(BaseModel):
    decision: str  # "genuine" or "flagged"
    remarks: str | None = None


class FeedbackResponse(BaseModel):
    id: int
    paper_id: int
    decision: str
    remarks: str | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)
